File: server/routes/watchlists.js
const express = require('express');
const router = express.Router();
const pool = require('../db');
const { authenticateToken } = require('../middleware/auth');

// Get user's watchlists
router.get('/', authenticateToken, async (req, res) => {
  try {
    const [watchlists] = await pool.query(`
      SELECT w.*, COUNT(wi.Movie_ID) AS item_count
      FROM Watchlist w
      LEFT JOIN Watchlist_Item wi ON w.Watchlist_ID = wi.Watchlist_ID
      WHERE w.User_ID = ?
      GROUP BY w.Watchlist_ID
      ORDER BY w.Created_at DESC
    `, [req.user.id]);
    res.json(watchlists);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Server error' });
  }
});

// Get watchlist with movies
router.get('/:id', authenticateToken, async (req, res) => {
  try {
    const [watchlist] = await pool.query(
      'SELECT * FROM Watchlist WHERE Watchlist_ID = ? AND User_ID = ?',
      [req.params.id, req.user.id]
    );
    if (watchlist.length === 0) return res.status(404).json({ error: 'Watchlist not found' });

    const [movies] = await pool.query(`
      SELECT m.*, ROUND(AVG(r.Rating_value), 1) AS avg_rating,
        GROUP_CONCAT(DISTINCT g.Genre_name SEPARATOR ', ') AS genres
      FROM Watchlist_Item wi
      JOIN Movie m ON wi.Movie_ID = m.Movie_ID
      LEFT JOIN Rating r ON m.Movie_ID = r.Movie_ID
      LEFT JOIN Movie_Genre mg ON m.Movie_ID = mg.Movie_ID
      LEFT JOIN Genre g ON mg.Genre_ID = g.Genre_ID
      WHERE wi.Watchlist_ID = ?
      GROUP BY m.Movie_ID
    `, [req.params.id]);

    res.json({ ...watchlist[0], movies });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Server error' });
  }
});

// Create watchlist
router.post('/', authenticateToken, async (req, res) => {
  try {
    const { name } = req.body;
    if (!name) return res.status(400).json({ error: 'Watchlist name required' });
    const [result] = await pool.query(
      'INSERT INTO Watchlist (Watchlist_name, User_ID) VALUES (?, ?)',
      [name, req.user.id]
    );
    res.status(201).json({ Watchlist_ID: result.insertId, Watchlist_name: name });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Server error' });
  }
});

// Add movie to watchlist
router.post('/:id/movies', authenticateToken, async (req, res) => {
  try {
    const { movieId } = req.body;
    if (!movieId) return res.status(400).json({ error: 'Movie ID required' });

    const [watchlist] = await pool.query(
      'SELECT * FROM Watchlist WHERE Watchlist_ID = ? AND User_ID = ?',
      [req.params.id, req.user.id]
    );
    if (watchlist.length === 0) return res.status(404).json({ error: 'Watchlist not found' });

    // Use INSERT IGNORE to prevent race conditions - silently ignore if already exists
    const [result] = await pool.query(
      'INSERT IGNORE INTO Watchlist_Item (Watchlist_ID, Movie_ID) VALUES (?, ?)',
      [req.params.id, movieId]
    );
    
    // Check if movie was actually inserted or already existed
    if (result.affectedRows === 0) {
      return res.status(409).json({ error: 'This movie is already added to this watchlist.' });
    }

    res.json({ message: 'Movie added to watchlist' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Server error' });
  }
});

// Remove movie from watchlist
router.delete('/:id/movies/:movieId', authenticateToken, async (req, res) => {
  try {
    await pool.query(
      'DELETE FROM Watchlist_Item WHERE Watchlist_ID = ? AND Movie_ID = ?',
      [req.params.id, req.params.movieId]
    );
    res.json({ message: 'Movie removed' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Server error' });
  }
});

module.exports = router;

File: backend-flask/app/routes/watchlists.py
from flask import Blueprint, jsonify, request
from flask_jwt_extended import get_jwt_identity, jwt_required
from sqlalchemy import text

from ..extensions import get_db_session

watchlists_bp = Blueprint('watchlists', __name__)


@watchlists_bp.get('')
@jwt_required()
def get_watchlists():
    user_id = int(get_jwt_identity())
    session = get_db_session()
    try:
        rows = session.execute(
            text(
                """
                SELECT w.*, COUNT(wi.Movie_ID) AS item_count
                FROM Watchlist w
                LEFT JOIN Watchlist_Item wi ON w.Watchlist_ID = wi.Watchlist_ID
                WHERE w.User_ID = :user_id
                GROUP BY w.Watchlist_ID
                ORDER BY w.Created_at DESC
                """
            ),
            {'user_id': user_id},
        ).mappings().all()
        return jsonify([dict(row) for row in rows])
    except Exception:
        return jsonify({'error': 'Server error'}), 500
    finally:
        session.close()


@watchlists_bp.get('/<int:watchlist_id>')
@jwt_required()
def get_watchlist(watchlist_id):
    user_id = int(get_jwt_identity())
    session = get_db_session()
    try:
        watchlist = session.execute(
            text('SELECT * FROM Watchlist WHERE Watchlist_ID = :watchlist_id AND User_ID = :user_id'),
            {'watchlist_id': watchlist_id, 'user_id': user_id},
        ).mappings().first()

        if not watchlist:
            return jsonify({'error': 'Watchlist not found'}), 404

        movies = session.execute(
            text(
                """
                SELECT m.*, ROUND(AVG(r.Rating_value), 1) AS avg_rating,
                    GROUP_CONCAT(DISTINCT g.Genre_name SEPARATOR ', ') AS genres
                FROM Watchlist_Item wi
                JOIN Movie m ON wi.Movie_ID = m.Movie_ID
                LEFT JOIN Rating r ON m.Movie_ID = r.Movie_ID
                LEFT JOIN Movie_Genre mg ON m.Movie_ID = mg.Movie_ID
                LEFT JOIN Genre g ON mg.Genre_ID = g.Genre_ID
                WHERE wi.Watchlist_ID = :watchlist_id
                GROUP BY m.Movie_ID
                """
            ),
            {'watchlist_id': watchlist_id},
        ).mappings().all()

        response = dict(watchlist)
        response['movies'] = [dict(row) for row in movies]
        return jsonify(response)
    except Exception:
        return jsonify({'error': 'Server error'}), 500
    finally:
        session.close()


@watchlists_bp.post('')
@jwt_required()
def create_watchlist():
    user_id = int(get_jwt_identity())
    payload = request.get_json(silent=True) or {}
    name = (payload.get('name') or '').strip()

    if not name:
        return jsonify({'error': 'Watchlist name required'}), 400

    session = get_db_session()
    try:
        result = session.execute(
            text('INSERT INTO Watchlist (Watchlist_name, User_ID) VALUES (:name, :user_id)'),
            {'name': name, 'user_id': user_id},
        )
        session.commit()
        return jsonify({'Watchlist_ID': result.lastrowid, 'Watchlist_name': name}), 201
    except Exception:
        session.rollback()
        return jsonify({'error': 'Server error'}), 500
    finally:
        session.close()


@watchlists_bp.post('/<int:watchlist_id>/movies')
@jwt_required()
def add_movie_to_watchlist(watchlist_id):
    user_id = int(get_jwt_identity())
    payload = request.get_json(silent=True) or {}
    movie_id = payload.get('movieId')

    if not movie_id:
        return jsonify({'error': 'Movie ID required'}), 400

    session = get_db_session()
    try:
        # ── Conflicting Transaction: Concurrent Watchlist Add ─────────────────
        # A user double-clicking "Add to Watchlist", or two open browser tabs,
        # could fire two POST requests simultaneously. Without locking, both
        # sessions read no existing Watchlist_Item row, then both attempt
        # INSERT — crashing on the composite PRIMARY KEY (Watchlist_ID, Movie_ID).
        # SELECT ... FOR UPDATE on the parent Watchlist row acts as a mutex:
        # Session B blocks here until Session A commits. When B resumes, it
        # re-checks Watchlist_Item, finds the row now exists, and cleanly
        # returns 409 instead of a database error.
        session.execute(text('SET TRANSACTION ISOLATION LEVEL REPEATABLE READ'))
        session.execute(text('START TRANSACTION'))

        watchlist = session.execute(
            text(
                'SELECT * FROM Watchlist'
                ' WHERE Watchlist_ID = :watchlist_id AND User_ID = :user_id FOR UPDATE'
            ),
            {'watchlist_id': watchlist_id, 'user_id': user_id},
        ).mappings().first()

        if not watchlist:
            session.rollback()
            return jsonify({'error': 'Watchlist not found'}), 404

        # Re-check inside the lock — this is the critical read that prevents
        # the duplicate insert under concurrent access.
        existing = session.execute(
            text(
                'SELECT 1 FROM Watchlist_Item'
                ' WHERE Watchlist_ID = :watchlist_id AND Movie_ID = :movie_id LIMIT 1'
            ),
            {'watchlist_id': watchlist_id, 'movie_id': movie_id},
        ).mappings().first()

        if existing:
            session.rollback()
            return jsonify({'error': 'This movie is already added to this watchlist.'}), 409

        session.execute(
            text('INSERT INTO Watchlist_Item (Watchlist_ID, Movie_ID) VALUES (:watchlist_id, :movie_id)'),
            {'watchlist_id': watchlist_id, 'movie_id': movie_id},
        )
        session.commit()
        return jsonify({'message': 'Movie added to watchlist'})
    except Exception:
        session.rollback()
        return jsonify({'error': 'Server error'}), 500
    finally:
        session.close()



@watchlists_bp.delete('/<int:watchlist_id>/movies/<int:movie_id>')
@jwt_required()
def remove_movie_from_watchlist(watchlist_id, movie_id):
    user_id = int(get_jwt_identity())
    session = get_db_session()
    try:
        watchlist = session.execute(
            text('SELECT * FROM Watchlist WHERE Watchlist_ID = :watchlist_id AND User_ID = :user_id'),
            {'watchlist_id': watchlist_id, 'user_id': user_id},
        ).mappings().first()

        if not watchlist:
            return jsonify({'error': 'Watchlist not found'}), 404

        session.execute(
            text('DELETE FROM Watchlist_Item WHERE Watchlist_ID = :watchlist_id AND Movie_ID = :movie_id'),
            {'watchlist_id': watchlist_id, 'movie_id': movie_id},
        )
        session.commit()
        return jsonify({'message': 'Movie removed'})
    except Exception:
        session.rollback()
        return jsonify({'error': 'Server error'}), 500
    finally:
        session.close()
